### PR TITLE
ELPP-2983 - profile page - login control

### PR DESCRIPTION
**Pull request by davidcmoulton**

### CODE REVIEW

**Comment by davidcmoulton** [comment]
@thewilkybarkid the assets still need deconflicting after the latest merges into master, but the PHP & tests are ready for review.

File: tests/src/ViewModel/ProfileLoginControlLoggedInTest.php [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     * @test
+     */
+    public function it_must_be_given_a_profile_home_link()
+    {
+        $this->expectException(InvalidArgumentException::class);
+
+        ProfileLoginControl::LoggedIn('', 'Display Name', $this->linkFieldRoots['input'], $this->linkFieldData['input']);
```
Comment (thewilkybarkid) [inline tests/src/ViewModel/ProfileLoginControlLoggedInTest.php]: While method names are case insensitive, better to match it.

File: src/ViewModel/ProfileLoginControl.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+namespace eLife\Patterns\ViewModel;
+
+use Assert\Assertion;
+use eLife\Patterns\ArrayAccessFromProperties;
+use eLife\Patterns\ArrayFromProperties;
+use eLife\Patterns\SimplifyAssets;
+use eLife\Patterns\ViewModel;
+use Traversable;
+
+final class ProfileLoginControl implements ViewModel
```
[the remark this answers is not in the document]
Comment (davidcmoulton) [reply src/ViewModel/ProfileLoginControl.php]: Good point. It originally had a load of profile-specific properties, but most of those went in the last reworking. Can probably make the remainder generic enough to lose the `Profile` name prefix.

File: src/ViewModel/ProfileLoginControl.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        return $loggedInControl;
+    }
+
+    public static function notLoggedIn(string $uri) : ProfileLoginControl
+    {
+        Assertion::notBlank($uri);
+
+        $notLoggedInControl = new static();
+        $notLoggedInControl->isLoggedIn = null;
+        $notLoggedInControl->button = Button::link('Log in / Register', $uri, Button::SIZE_EXTRA_SMALL, Button::STYLE_CONFIRM);
```
[the remark this answers is not in the document]
Comment (davidcmoulton) [reply src/ViewModel/ProfileLoginControl.php]: :+1:  for it as data. It would make any i18n easier.

**Comment by davidcmoulton** [comment]
@thewilkybarkid changes made as requested.

File: tests/src/ViewModel/LoginControlLoggedInTest.php [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        $this->expectException(InvalidArgumentException::class);
+
+        LoginControl::loggedIn(
+            '/defaultUri',
+            'Display Name',
+            [
+                'root-name-1' => [
+                    'uri' => '/uriOne',
+                    'text' => 'Text one',
```
Comment (thewilkybarkid) [inline tests/src/ViewModel/LoginControlLoggedInTest.php]: There's no need to require an ID like `root-name-1`. The input can just be `['Text one' => '/uriOne']`.

File: src/ViewModel/LoginControl.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Assertion::notBlank($text);
+
+        $notLoggedInControl = new static();
+        $notLoggedInControl->isLoggedIn = null;
+        $notLoggedInControl->button = Button::link($text, $uri, Button::SIZE_EXTRA_SMALL, Button::STYLE_CONFIRM);
+
+        return $notLoggedInControl;
+    }
+
+    private static function buildLinkFieldsDataAttributeValues(array $linkFields)
```
Comment (thewilkybarkid) [inline src/ViewModel/LoginControl.php]: Not being used statically.

File: tests/src/ViewModel/LoginControlLoggedInTest.php [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+        $profileLoginControl = LoginControl::loggedIn($data['defaultUri'], $data['displayName'], $this->linkFields['input']);
+
+        $this->assertSame($data['isLoggedIn'], $profileLoginControl['isLoggedIn']);
+        $this->assertSame($data['defaultUri'], $profileLoginControl['defaultUri']);
+
+        $this->assertSame($this->linkFields['expectedOutput']['linkFieldRoots'], $profileLoginControl['linkFieldRoots']);
+        $this->assertSame($this->linkFields['expectedOutput']['linkFieldData'], $profileLoginControl['linkFieldData']);
+
+        // Ugly hack to get the properties in the correct order for the following assertSame :-(
```
Comment (thewilkybarkid) [inline tests/src/ViewModel/LoginControlLoggedInTest.php]: Use `assertSameWithoutOrder` (or whatever it's called).

File: tests/src/ViewModel/SiteHeaderTest.php [hunk 7]
```diff
@@ -119,7 +121,8 @@ public function viewModelProvider() : array
             [
                 NavLinkedItem::asIcon(new Link('text-first', '/path/first'), $img),
                 NavLinkedItem::asLink(new Link('text-first', '/path/first'), false),
-                NavLinkedItem::asButton(Button::link('button text', '/button/path')),
+                NavLinkedItem::asButton(Button::link('button text', '/button/path')),                NavLinkedItem::asLoginControl(LoginControl::notLoggedIn('Log in / Register', '/log-in')),
+
```
Comment (thewilkybarkid) [inline tests/src/ViewModel/SiteHeaderTest.php]: Item should be on this line

File: src/ViewModel/LoginControl.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        $loggedInControl->isLoggedIn = true;
+        $loggedInControl->displayName = $displayName;
+        $loggedInControl->subsidiaryText = $subsidiaryText;
+        $loggedInControl->defaultUri = $defaultUri;
+        $loggedInControl->icon = $icon;
+
+        if (!empty($linkFields)) {
+            foreach ($linkFields as $text => $uri) {
+                Assertion::notBlank($text);
+                Assertion::notBlank($uri);
```
Comment (thewilkybarkid) [inline src/ViewModel/LoginControl.php]: Should be able to use `Assertion::allNotBlank()`.

**Pull request merged**